torch/csrc/autograd/engine.cpp

```diff
@@ -321,60 +317,52 @@ auto Engine::thread_main(
       // Scope this block of execution since NodeTask is not needed after this
       // block and can be deallocated (release any references to grad tensors
       // as part of inputs_).
-      NodeTask task = queue->pop();
+      auto task = local_ready_queue->pop();
       // This will only work if the worker is running a non backward task
       // TODO Needs to be fixed this to work in all cases
-      if (task.isShutdownTask_) {
+      if (task->isShutdownTask_) {
         C10_LOG_API_USAGE_ONCE("torch.autograd.thread_shutdown");
         break;
       }
-
-      if (!(local_graph_task = task.base_.lock())) {
-        // GraphTask for function is no longer valid, skipping further
-        // execution.
+      local_graph_task = execute_node_task(task);
+      if (!local_graph_task) {
+        // GraphTask for function is no longer valid, skipping further execution.
         continue;
       }
-
-      if (task.fn_ && !local_graph_task->has_error_.load()) {
-        AutoGradMode grad_mode(local_graph_task->grad_mode_);
-        try {
-          evaluate_function(local_graph_task, task.fn_.get(), task.inputs_);
-        } catch (std::exception& e) {
-          thread_on_exception(local_graph_task, task.fn_, e);
-        }
-      }
     }
 
-    // Decrement the outstanding tasks.
-    --local_graph_task->outstanding_tasks_;
-
     // Check if we've completed execution.
     bool gt_completed = graph_task_completed(local_graph_task);
     if (gt_completed) {
       // We don't need to explicitly notify the owner thread, since
       // 'mark_graph_task_completed' would mark the Future as completed and this
       // would notify the owner thread that the task has been completed.
       mark_graph_task_completed(local_graph_task);
+
+      // The CPU worker thread is actually the thread that initially requested
+      // the autograd computation. Now that we're done, we need to break out of
+      // the worker loop so we can continue executing the rest of the calling code!
+      if (worker_device == CPU_DEVICE) {
+        break;
```

Comment:
Wouldn't we break out of the loop naturally since the while condition would no longer be true? Also, don't we need to execute the logic below to notify the owning thread?